migrator.py

```diff
@@ -3,39 +3,153 @@
 import configparser
 import logging
 import os
+import traceback
+import xml.etree.ElementTree as ET
+
 from pathlib import Path
 from typing import Dict, List, Optional, Set
 
 from digikam_db import DigikamDb
 import rect64
 
 _PICASA_INI_FILE = '.picasa.ini'
+_OLD_PICASA_INI_FILE = 'Picasa.ini' # found this in folders from 2002!
 _PICASA_TAG_NAME = 'Picasa'
 _UNKNOWN_FACE_ID = 'ffffffffffffffff'
 _FACE_TAG_REGION_PROPERTY = 'tagRegion'
 
 ContactTags = Dict[str, Optional[int]]
-
+GlobalNames = Dict[str, str] # maps contact_id to name globally (but may be overridden per directory)
 
 def _is_photo_file(file: str) -> bool:
     file = file.lower()
     return file.endswith('.jpg') or file.endswith('.jpeg') or file.endswith('.raw') or file.endswith('.psd')
 
+def learn_contact_ids(input_dir: Path, ini_file_name: str, contact_id_2_nameset: Dict[str, set]):
+    """ Learn contact_id to name mapping from [Contacts2] sections of .ini file"""
+    # Read ini file.
+    #logging.info("learn_contact_ids: input_dir=%s" % input_dir)
+    ini = configparser.ConfigParser(strict=False)
+    ini_file = input_dir / Path(ini_file_name)
+    ini.read(ini_file, encoding='utf8')
+    try:
+        for contact_id, value in ini['Contacts2'].items():
+            person_name = value.split(';')[0]
+            if not contact_id in contact_id_2_nameset:
+                logging.info(f"Learned name for {contact_id}='{person_name}' from {ini_file}")
+                contact_id_2_nameset[contact_id]=set([person_name])
+            else:
+                if person_name not in contact_id_2_nameset[contact_id]:
+                     logging.warning(f"Learned additional name for {contact_id}='{person_name}' from {ini_file}")
+                contact_id_2_nameset[contact_id].add(person_name)
+    except KeyError:
+        # likely no Contacts2 section
+        pass
+    except Exception as e:
+        logging.warning("Exception while learning contact ids")
+        logging.exception(str(e))
+
+def learn_old_contact_ids(input_dir: Path, ini_file_name: str, contact_id_2_nameset: Dict[str, set]):
+    """ Learn contact_id to name mapping from old [Contacts] sections of .ini file"""
+    # Prerequisite: Must have learned all that is possible from new [Contacts2] already
```

Comment:
How about merging the two functions? Then the function can fulfill this prerequisite by just running things in the correct order internally, and we only need to parse the file once.

Reply:
Merged